types/block.go

```diff
@@ -917,6 +918,32 @@ func (commit *Commit) ValidateBasic() error {
 	return nil
 }
 
+// MedianTime computes the median time for a Commit based on the associated validator set.
+// The median time is the weighted median of the Timestamp fields of the commit votes,
+// with heights defined by the validator's voting powers.
+// The BFT Time algorithm ensures that the computed median time is always picked among
+// the timestamps produced by honest processes, i.e., faulty processes cannot arbitrarily
+// increase or decrease the median time.
+// See: https://github.com/cometbft/cometbft/blob/main/spec/consensus/bft-time.md
+func (commit *Commit) MedianTime(validators *ValidatorSet) time.Time {
+	weightedTimes := make([]*cmttime.WeightedTime, len(commit.Signatures))
```

Comment:
Another possible implementation of this `MedianTime` function would be using a vote set, might simplify a few things

```
func (commit *Commit) MedianTime(validators *ValidatorSet) time.Time {

	weightedTimes := make([]*cmttime.WeightedTime, len(commit.Signatures))
	totalVotingPower := int64(0)

	voteSet := commit.ToVoteSet("", validators)
	for i, v := range voteSet.votes {
		if v.CommitSig().BlockIDFlag != BlockIDFlagAbsent {
			_, val := validators.GetByIndex(v.ValidatorIndex)
			if val != nil {
				totalVotingPower += val.VotingPower
				weightedTimes[i] = cmttime.NewWeightedTime(v.Timestamp, val.VotingPower)
			}
		}
	}
	return cmttime.WeightedMedian(weightedTimes, totalVotingPower)
}
```

Reply:
This is actually a good question. Is this method indeed computing the median of all non-nil values?

Reply:
By the way:

```
// TODO: find which commit removed this and make sure it's in our list
// WeightedTime for computing a median.
type WeightedTime struct {
        Time   time.Time
        Weight int64
}
```

Same file as `WeightedMedian` is implemented.

Reply:
We could just append all non-nil entries to the slice, instead of pre-allocating it.

Reply:
> Is this method indeed computing the median of all non-nil values?

It is, but this is not tested. I've run some tests here and it works as expected.

Reply:
Thanks for the suggestions @andynog.
The point of this PR was to move the function to a place where it is accessible for other packages. So I'm not inclined to change a battle-tested implementation here.
If you think your suggestions may greatly improve performance, code elegance, maintainability, etc. feel free to open an issue and we can tackle it when this branch lands on `main`. Here we're in the critical path and I'd like to avoid blocking this PR unless there is a concern.